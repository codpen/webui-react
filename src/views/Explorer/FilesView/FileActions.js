import React from "react";
import {Button, DropdownItem, DropdownMenu, DropdownToggle, UncontrolledButtonDropdown} from "reactstrap";
import * as PropTypes from "prop-types";
import * as RclonePropTypes from "../../../utils/RclonePropTypes";

function FileActions({downloadHandle, deleteHandle, item, linkShareHandle}) {
    const confirmDelete = (deleteHandle, item) => {
        if (window.confirm(`Are you sure you want to delete ${item.Name}`)) {
            deleteHandle(item);
        }
    }

    const {IsDir} = item;
    // let {ID, Name} = item;
    // // Using fallback as fileName when the ID is not available (for local file system)
    // if (ID === undefined) {
    //     ID = Name;
    // }

    return (
        <div data-test="fileActionsComponent">
            {!IsDir && <Button color="link" onClick={() => downloadHandle(item)} data-test="btn-download">
                <i className={"fa fa-cloud-download fa-lg d-inline"}/>
            </Button>}
            <Button color="link">
                <i className="fa fa-info-circle"/>
            </Button>

            <UncontrolledButtonDropdown>
                <DropdownToggle color="link">
                    <i className="fa fa-ellipsis-v"/>
                </DropdownToggle>
                <DropdownMenu>
                    <DropdownItem header>Actions</DropdownItem>
                    <DropdownItem data-test="btn-share-with-link" onClick={() => linkShareHandle(item)}><i
                        className="fa fa-share fa-lg d-inline"/> Share with link</DropdownItem>
                    <DropdownItem divider/>
                    <DropdownItem data-test="btn-delete-item" onClick={() => confirmDelete(deleteHandle, item)}><i
                        className="fa fa-remove fa-lg d-inline text-danger"/> Delete </DropdownItem>
                </DropdownMenu>
            </UncontrolledButtonDropdown>
        </div>
    )
}

FileActions.propTypes = {
    downloadHandle: PropTypes.func.isRequired,
    deleteHandle: PropTypes.func.isRequired,
    item: RclonePropTypes.PROP_ITEM.isRequired,
    linkShareHandle: PropTypes.func.isRequired

}

export default FileActions;