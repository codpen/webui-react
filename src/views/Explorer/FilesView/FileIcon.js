import React from "react";
import * as PropTypes from "prop-types";

const mimeClassMap = {
    "application/pdf": "fa-file-pdf-o",
    "image/jpeg": "fa-file-image-o",
    "application/rar": "fa-file-archive-o",
    "application/x-rar-compressed": "fa-file-archive-o",
    "application/zip": "fa-file-archive-o",
    "text/plain": "fa-file-text-o",
    "text/x-vcard": "fa-address-card-o"
}

function FileIcon({IsDir, MimeType}) {
    let className = mimeClassMap[MimeType];
    if (IsDir) {
        className = "fa-folder"
    }
    if (!className) className = "fa-file";

    return <i className={className + " fa fa-lg"} data-test="fileIconComponent"/>;
}

FileIcon.propTypes = {
    IsDir: PropTypes.bool.isRequired,
    MimeType: PropTypes.string.isRequired
}

export default FileIcon;