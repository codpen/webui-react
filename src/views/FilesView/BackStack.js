class BackStack {
    constructor() {
        this.items = [];
        this.count = 0;
    }

    getLength() {
        return this.count;
    }

    push(item) {
        this.items.push(item);
        this.count = this.count + 1;
    }

    pop() {
        if (this.count > 0) {
            this.count = this.count - 1;
        }

        return this.items.pop();
    }

    peek() {
        return this.items.slice(-1)[0];
    }

    empty() {
        this.items = [];
        this.count = 0;
    }
}

export default BackStack;