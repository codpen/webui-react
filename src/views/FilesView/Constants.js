export const ItemTypes = {
    FILECOMPONENT: 'FileComponent'
}