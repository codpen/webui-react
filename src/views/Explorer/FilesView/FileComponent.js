import React from "react";
import {
    Button,
    Card,
    CardBody,
    CardFooter,
    DropdownItem,
    DropdownMenu,
    DropdownToggle,
    UncontrolledButtonDropdown
} from "reactstrap";

import {ItemTypes} from './Constants'
import {DragSource} from 'react-dnd'
import {formatBytes} from "../../../utils/Tools";
import {performCopyFile, performMoveFile} from "../../../utils/API/API";
import {toast} from "react-toastify";
import * as PropTypes from "prop-types";
import handleViewport from 'react-in-viewport';
import MediaWidget, {isMedia} from "../../Base/MediaWidget/MediaWidget";


const fileComponentSource = {
    beginDrag(props) {
        // console.log("props", props, props.remoteName);
        const {Name, Path, IsDir} = props.item;
        return {
            Name: Name, Path: Path, IsDir: IsDir, remoteName: props.remoteName, remotePath: props.remotePath
        }
    },

    async endDrag(props, monitor, component) {
        // console.log("EndDrag", monitor.getDropResult());
        console.log(props, "Component:", component);
        try {
            if (monitor.getDropResult() && component) {


                const {srcRemoteName, srcRemotePath, destRemoteName, destRemotePath, Name, IsDir, dropEffect, updateHandler} = monitor.getDropResult();

                if (dropEffect === "move") { /*Default operation without holding alt is copy, named as move in react-dnd*/
                    // if (component.props.canCopy) {
                    let res = await performCopyFile(srcRemoteName, srcRemotePath, destRemoteName, destRemotePath, Name, IsDir);
                    console.log("Copy", res);
                    updateHandler();
                    if (IsDir) {
                        toast.info(`Directory copying started in background: ${Name}`);
                    } else {
                        toast.info(`File copying started in background: ${Name}`);
                    }
                    // } else {
                    //     toast.error("This remote does not support copying");
                    // }

                } else {
                    // if (component.props.canMove) {
                    await performMoveFile(srcRemoteName, srcRemotePath, destRemoteName, destRemotePath, Name, IsDir);
                    updateHandler();
                    if (IsDir) {
                        toast.info(`Directory moving started in background: ${Name}`);
                    } else {
                        toast.info(`Directory moving started in background: ${Name}`);
                    }
                    // } else {
                    //     toast.error("This remote does not support moving");
                    // }

                }
            }
        } catch (e) {
            const error = e.response ? e.response : e;
            console.log(JSON.stringify(error));

            toast.error(`Error copying file(s). ${error}`, {
                autoClose: false
            });

        }
    }
};

function collect(connect, monitor) {
    return {
        connectDragSource: connect.dragSource(),
        connectDragPreview: connect.dragPreview(),
        isDragging: monitor.isDragging()
    }
}

function FileIcon({IsDir, MimeType}, ...props) {
    let className = "fa-file";
    if (IsDir) {
        className = "fa-folder";
    } else if (MimeType === "application/pdf") {
        className = "fa-file-pdf-o";
    } else if (MimeType === "image/jpeg") {
        className = "fa-file-image-o";
    } else if (MimeType === "application/rar" || MimeType === "application/x-rar-compressed" || MimeType === " application/zip") {
        className = "fa-file-archive-o";
    } else if (MimeType === "text/plain") {
        className = "fa-file-text-o";
    } else if (MimeType === "text/x-vcard") {
        className = "fa-address-card-o";
    }
    return <i className={className + " fa fa-lg"}/>;
}

function confirmDelete(deleteHandle, item) {
    if (window.confirm(`Are you sure you want to delete ${item.Name}`)) {
        deleteHandle(item);
    }
}

function Actions({downloadHandle, deleteHandle, item, linkShareHandle}) {

    const {IsDir} = item;
    // let {ID, Name} = item;
    // // Using fallback as fileName when the ID is not available (for local file system)
    // if (ID === undefined) {
    //     ID = Name;
    // }


    if (!IsDir) {

        return (
            <React.Fragment>
                <Button color="link" onClick={() => downloadHandle(item)}>
                    <i className={"fa fa-cloud-download fa-lg d-inline"}/>
                </Button>
                <Button color="link">
                    <i className="fa fa-info-circle"/>
                </Button>
                {/* TODO: Find a way to make this work*/}
                {/*<UncontrolledTooltip placement="right" target={"#tooltip"+ID}>*/}
                {/*    {item}*/}
                {/*</UncontrolledTooltip>*/}

                <UncontrolledButtonDropdown>
                    <DropdownToggle color="link">
                        <i className="fa fa-ellipsis-v"/>
                    </DropdownToggle>
                    <DropdownMenu>
                        <DropdownItem header>Actions</DropdownItem>
                        <DropdownItem onClick={() => linkShareHandle(item)}><i
                            className="fa fa-share fa-lg d-inline"/> Share with link</DropdownItem>
                        <DropdownItem divider/>
                        <DropdownItem onClick={() => confirmDelete(deleteHandle, item)}><i
                            className="fa fa-remove fa-lg d-inline text-danger"/> Delete </DropdownItem>
                    </DropdownMenu>
                </UncontrolledButtonDropdown>
            </React.Fragment>

        );
    } else {
        return (
            <React.Fragment>

                <UncontrolledButtonDropdown>
                    <DropdownToggle color="link">
                        <i className="fa fa-ellipsis-v"/>
                    </DropdownToggle>
                    <DropdownMenu>
                        <DropdownItem header>Actions</DropdownItem>
                        <DropdownItem onClick={() => linkShareHandle(item)}><i
                            className="fa fa-share fa-lg d-inline"/> Share with link</DropdownItem>
                        <DropdownItem divider/>
                        <DropdownItem onClick={() => confirmDelete(deleteHandle, item)}><i
                            className="fa fa-remove fa-lg d-inline text-danger"/> Delete </DropdownItem>
                    </DropdownMenu>
                </UncontrolledButtonDropdown>
            </React.Fragment>
        )
    }
}

// Non used props are required for drag-and-drop functionality
class FileComponent extends React.Component {

    /*
    MimeTypes: https://www.freeformatter.com/mime-types-list.html
    * {
    * For Directory
			"ID": "18DsZ4ne6XV3qwDZQCBj2nAEwouFMxudB",
			"IsDir": true,
			"MimeType": "inode/directory",
			"ModTime": "2019-02-12T14:23:33.440Z",
			"Name": "two pass 28-1-19",
			"Path": "two pass 28-1-19",
			"Size": -1
		},
		*
		* // For non-directory
		* {
			"ID": "1u4D6-UdxhJYY8AVd8FcTN2Tl73W1RXsk",
			"IsDir": false,
			"MimeType": "application/octet-stream",
			"ModTime": "2018-11-18T13:14:54.068Z",
			"Name": "streamlined-gdoc.gdoc",
			"Path": "streamlined-gdoc.gdoc",
			"Size": 173
		},

    * */
    render() {
        const {containerID, inViewport, item, loadImages, clickHandler, downloadHandle, linkShareHandle, deleteHandle, connectDragSource, gridMode, itemIdx /*isDragging, remoteName*/} = this.props;

        const {IsDir, MimeType, ModTime, Name, Size} = item;


        let modTime = new Date(Date.parse(ModTime));

        if (gridMode === "card") {
            return connectDragSource(
                <div className={IsDir ? "" : "col-md-4"}>
                    <Card>
                        <CardBody onClick={(e) => clickHandler(e, item)}>

                            {loadImages && isMedia(MimeType) ?
                                <MediaWidget containerID={containerID} item={item} inViewport={inViewport}/> :
                                <FileIcon IsDir={IsDir} MimeType={MimeType}/>
                            }


                            {Name}
                        </CardBody>
                        <CardFooter>
                            <Actions downloadHandle={downloadHandle} linkShareHandle={linkShareHandle}
                                     deleteHandle={deleteHandle} item={item}/>
                        </CardFooter>
                    </Card>
                </div>
            )
        } else {
            return connectDragSource(
                <tr className={"pointer-cursor"}>
                    <td className="d-none d-md-table-cell"><input type="checkbox"/></td>
                    <td onClick={(e) => clickHandler(e, item)} id={"file" + itemIdx}>
                        <FileIcon IsDir={IsDir} MimeType={MimeType}/> {Name}


                    </td>
                    <td>{Size === -1 ? "-" : formatBytes(Size, 2)}</td>
                    <td className="d-none d-md-table-cell">{modTime.toLocaleDateString()}</td>
                    <td><Actions downloadHandle={downloadHandle} linkShareHandle={linkShareHandle}
                                 deleteHandle={deleteHandle} item={item}/></td>
                </tr>
            )
        }
    }
}


FileComponent.propTypes = {
    item: PropTypes.object.isRequired,
    clickHandler: PropTypes.func.isRequired,
    downloadHandle: PropTypes.func.isRequired,
    deleteHandle: PropTypes.func.isRequired,
    linkShareHandle: PropTypes.func.isRequired,
    remoteName: PropTypes.string.isRequired,
    remotePath: PropTypes.string.isRequired,
    gridMode: PropTypes.string,
    containerID: PropTypes.string.isRequired,
    canMove: PropTypes.bool.isRequired,
    canCopy: PropTypes.bool.isRequired,
    loadImages: PropTypes.bool.isRequired,
    isBucketBased: PropTypes.bool.isRequired

};

const MyViewPort = handleViewport(FileComponent, {rootMargin: '-1.0px'});

export default DragSource(ItemTypes.FILECOMPONENT, fileComponentSource, collect)(MyViewPort);
