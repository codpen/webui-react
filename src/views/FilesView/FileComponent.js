import React from "react";
import {Button} from "reactstrap";

import {ItemTypes} from './Constants'
import {DragSource} from 'react-dnd'
import {formatBytes} from "../../utils/Tools";

const fileComponentSource = {
    beginDrag(props) {
        console.log("props", props, props.remoteName);
        const {Name, Path, IsDir} = props.item;
        return {
            Name: Name, Path: Path, IsDir: IsDir, remoteName: props.remoteName
        }
    }
};

function collect(connect, monitor) {
    return {
        connectDragSource: connect.dragSource(),
        connectDragPreview: connect.dragPreview(),
        isDragging: monitor.isDragging()
    }
}


// TODO: Add mode parameter for card view or list view
function FileIcon({IsDir, MimeType}) {
    let className = "fa-file";
    if (IsDir) {
        className = "fa-folder";
    } else if (MimeType === "application/pdf") {
        className = "fa-file-pdf-o";
    } else if (MimeType === "image/jpeg") {
        className = "fa-file-image-o";
    } else if (MimeType === "application/rar" || MimeType === "application/x-rar-compressed" || MimeType === " application/zip") {
        className = "fa-file-archive-o";
    } else if (MimeType === "text/plain") {
        className = "fa-file-text-o";
    } else if (MimeType === "text/x-vcard") {
        className = "fa-address-card-o";
    }
    return <i className={className + " fa fa-lg"}/>;
}

function confirmDelete(deleteHandle, item) {
    if (window.confirm(`Are you sure you want to delete ${item.Name}`)) {
        deleteHandle(item);
    }
}

function Actions({downloadHandle, deleteHandle, IsDir, item}) {
    if (!IsDir) {
        return (
            <React.Fragment>
                <Button color="link" onClick={() => downloadHandle(item)}>
                    <i className={"fa fa-cloud-download fa-lg d-inline"}/>
                </Button>
                <Button color="link" onClick={() => confirmDelete(deleteHandle, item)}>
                    <i className={"fa fa-remove fa-lg d-inline"}/>
                </Button>
            </React.Fragment>

        );
    } else {
        return null;
    }
}

// Non used props are required for drag-and-drop functionality
function FileComponent({item, clickHandler, downloadHandle, deleteHandle, connectDragSource, isDragging, remoteName}) {
    /*
    MimeTypes: https://www.freeformatter.com/mime-types-list.html
    * {
    * For Directory
			"ID": "18DsZ4ne6XV3qwDZQCBj2nAEwouFMxudB",
			"IsDir": true,
			"MimeType": "inode/directory",
			"ModTime": "2019-02-12T14:23:33.440Z",
			"Name": "two pass 28-1-19",
			"Path": "two pass 28-1-19",
			"Size": -1
		},
		*
		* // For non-directory
		* {
			"ID": "1u4D6-UdxhJYY8AVd8FcTN2Tl73W1RXsk",
			"IsDir": false,
			"MimeType": "application/octet-stream",
			"ModTime": "2018-11-18T13:14:54.068Z",
			"Name": "streamlined-gdoc.gdoc",
			"Path": "streamlined-gdoc.gdoc",
			"Size": 173
		},

    * */
    const {IsDir, MimeType, ModTime, Name, Size} = item;

    return connectDragSource(
        <tr className={"pointer-cursor"}>
            <td><input type="checkbox"/></td>
            <th onClick={(e) => clickHandler(e, item)}><FileIcon IsDir={IsDir} MimeType={MimeType}/> {Name}</th>
            <td>{Size === -1 ? "NA" : formatBytes(Size, 2)}</td>
            {/*TODO: change the time format to required time using timezone as well*/}
            <td>{ModTime}</td>
            <td><Actions IsDir={IsDir} downloadHandle={downloadHandle} deleteHandle={deleteHandle} item={item}/></td>
        </tr>
    )
}

export default DragSource(ItemTypes.FILECOMPONENT, fileComponentSource, collect)(FileComponent);
