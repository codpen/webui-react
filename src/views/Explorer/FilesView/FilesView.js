import React from "react";
import axiosInstance from "../../../utils/API/API";
import {Alert, Button, Col, Container, Row, Spinner, Table} from "reactstrap";
import {DropTarget} from "react-dnd";
import FileComponent from "./FileComponent";
import {ItemTypes} from "./Constants";
import {toast} from "react-toastify";
import {addColonAtLast, changeListVisibility, changeSearchFilter, isEmpty} from "../../../utils/Tools";
import {connect} from "react-redux";
import {getFiles} from "../../../actions/explorerActions";
import {compose} from "redux";
import {changePath, navigateUp} from "../../../actions/explorerStateActions";
import LinkShareModal from "../../Base/LinkShareModal/LinkShareModal";
import ScrollableDiv from "../../Base/ScrollableDiv/ScrollableDiv";
import {FILES_VIEW_HEIGHT} from "../../../utils/Constants";
import {PROP_CURRENT_PATH, PROP_FS_INFO} from "../../../utils/RclonePropTypes";
import * as PropTypes from 'prop-types';
import ErrorBoundary from "../../../ErrorHandling/ErrorBoundary";
import urls from "../../../utils/API/endpoint";


/*
* Start code for react DND
* */

const filesTarget = {
    drop(props, monitor, component) {
        if (monitor.didDrop()) return;
        // console.log("drop", props, monitor, monitor.getItem(), component);

        let {Name, Path, IsDir, remoteName} = monitor.getItem();

        let srcRemoteName = addColonAtLast(remoteName);
        let srcRemotePath = Path;
        let destRemoteName = addColonAtLast(props.currentPath.remoteName);
        let destRemotePath = props.currentPath.remotePath;

        // console.log("drop:this", this);

        return {
            srcRemoteName,
            srcRemotePath,
            destRemoteName,
            destRemotePath,
            Name,
            IsDir,
            updateHandler: component.updateHandler
        }

    },
    canDrop(props, monitor) {
        const {remoteName, remotePath} = monitor.getItem();
        console.log(remoteName, props.currentPath.remoteName, remotePath, props.currentPath.remotePath);
        const destRemoteName = props.currentPath.remoteName;
        const destRemotePath = props.currentPath.remotePath;
        if (destRemoteName === remoteName) {
            return destRemotePath !== remotePath;
        }
        return true;
    }
};

function collect(connect, monitor) {
    return {
        connectDropTarget: connect.dropTarget(),
        isOver: monitor.isOver(),
        canDrop: monitor.canDrop()
    }
}

function renderOverlay() {
    return (
        <div
            style={{
                position: 'absolute',
                top: 0,
                left: 0,
                height: '100%',
                width: '100%',
                zIndex: 1,
                opacity: 0.5,
                backgroundColor: 'gray',
            }}
        />
    );
}

/*
* END code for react DND
* */

// Provides the up button view in the files view
function UpButtonComponent({upButtonHandle, gridMode}) {
    if (gridMode === "card") {
        return (
            <Col lg={12}>
                <Button onClick={() => upButtonHandle()}>Go Up</Button>
            </Col>
        )
    } else {
        return (
            <tr onClick={() => upButtonHandle()} className={"pointer-cursor"}>
                <td colSpan={1}/>
                <td colSpan={4}><i className={"fa fa-file-o"}/> Go Up...</td>
            </tr>);
    }
}

/**
 * FilesView component renders files in the file explorer.
 */
class FilesView extends React.PureComponent {


    constructor(props) {
        super(props);
        this.state = {
            isLoading: false,
            isDownloadProgress: false,
            downloadingItems: 0,
            shouldUpdate: true,
            showLinkShareModal: false,
            generatedLink: "",

        };
        this.handleFileClick = this.handleFileClick.bind(this);
        this.downloadHandle = this.downloadHandle.bind(this);
        this.deleteHandle = this.deleteHandle.bind(this);
    }

    closeLinkShareModal = () => {
        this.setState({
            showLinkShareModal: false
        })
    };

    showLinkShareModal = () => {
        this.setState({
            showLinkShareModal: true

        })
    };


    handleFileClick(e, item) {
        const {Path, IsDir, IsBucket} = item;
        if (IsDir || IsBucket) {
            this.updateRemotePath(Path, IsDir, IsBucket);
        } else {
            this.downloadHandle(item);
        }

    }

    updateRemotePath(newRemotePath, IsDir, IsBucket) {
        const {remoteName} = this.props.currentPath;

        let updateRemoteName = "";
        let updateRemotePath = "";


        if (IsBucket) {
            updateRemoteName = addColonAtLast(remoteName) + newRemotePath;
            updateRemotePath = "";
            // backStack.push({remoteName: addColonAtLast(backStack.peek().remoteName) + remotePath, remotePath: ""});

        } else if (IsDir) {
            updateRemoteName = remoteName;
            updateRemotePath = newRemotePath;
            // backStack.push({remoteName: backStack.peek().remoteName, remotePath: remotePath});
        }
        this.props.changePath(this.props.containerID, updateRemoteName, updateRemotePath);
    }


    getFilesList(showLoading = true) {
        const {remoteName, remotePath} = this.props.currentPath;

        this.props.getFiles(remoteName, remotePath);

    }

    async downloadHandle(item) {
        // let {remoteName, remotePath} = this.props;
        let {remoteName, remotePath} = this.props.currentPath;
        const {fsInfo} = this.props;
        let downloadUrl = "";
        if (fsInfo.Features.BucketBased) {
            downloadUrl = `/[${remoteName}]/${remotePath}/${item.Name}`;

        } else {
            downloadUrl = `/[${remoteName}:${remotePath}]/${item.Name}`;
        }

        this.setState((prevState) => {
            return {
                downloadingItems: prevState.downloadingItems + 1,
                isDownloadProgress: true
            };
        });

        let response = await axiosInstance({
            url: downloadUrl,
            method: 'GET',
            responseType: 'blob',
        });

        const url = window.URL.createObjectURL(new Blob([response.data]));
        const link = document.createElement('a');
        link.href = url;
        link.setAttribute('download', item.Name);
        document.body.appendChild(link);
        link.click();

        this.setState((prevState) => {
            return {
                downloadingItems: prevState.downloadingItems - 1,
            };
        }, () => {
            if (this.state.downloadingItems === 0) {
                this.setState({isDownloadProgress: false})
            }
        });
    }

    async deleteHandle(item) {
        let {remoteName} = this.props.currentPath;

        const data = {
            fs: addColonAtLast(remoteName),
            remote: item.Path,
        };
        try {
            if (item.IsDir) {

                await axiosInstance.post(urls.purge, data);

                this.updateHandler();
                toast.info(`${item.Name} deleted.`);

            } else {

                await axiosInstance.post(urls.deleteFile, data);
                this.updateHandler();
                toast.info(`${item.Name} deleted.`, {
                    autoClose: false
                });
            }
        } catch (e) {
            // console.log(`Error in deleting file`);
            toast.error(`Error deleting file. ${e}`, {
                autoClose: false
            });
        }

    }

    updateHandler = () => {

        const {remoteName, remotePath} = this.props.currentPath;
        this.getFilesList(remoteName, remotePath);
    };

    dismissAlert = (e) => {
        this.setState({isDownloadProgress: false});
    };

    linkShareHandle = (item) => {
        const {fsInfo} = this.props;
        if (fsInfo.Features.PublicLink) {
            console.log("Sharing link" + item.Name);
            const {remoteName} = this.props.currentPath;
            axiosInstance.post(urls.createPublicLink, {
                fs: addColonAtLast(remoteName),
                remote: item.Path
            }).then((res) => {
                // console.log("Public Link: " + res.data.url);

                this.setState({
                    generatedLink: res.data.url,
                    showLinkShareModal: true
                })
            }, (error) => {
                toast.error("Error Generating link: " + error)
            })
        } else {
            toast.error("This remote does not support public link");
        }

    };

    getFileComponents = (isDir) => {
        const {files, containerID, gridMode, fsInfo, loadImages} = this.props;
        const {remoteName, remotePath} = this.props.currentPath;
        // console.log(fsInfo, files);
        if (fsInfo && !isEmpty(fsInfo)) {
            return files.map((item, idx) => {
                let {ID, Name} = item;
                // Using fallback as fileName when the ID is not available (for local file system)
                if (ID === undefined) {
                    ID = Name;
                }
                if (item.IsDir === isDir) {
                    return (
                        <React.Fragment key={ID}>
                            <FileComponent item={item} clickHandler={this.handleFileClick}
                                           downloadHandle={this.downloadHandle} deleteHandle={this.deleteHandle}
                                           remoteName={remoteName} remotePath={remotePath} gridMode={gridMode}
                                           containerID={containerID}
                                           linkShareHandle={this.linkShareHandle}
                                           loadImages={loadImages}
                                           isBucketBased={fsInfo.Features.BucketBased}
                                           canCopy={fsInfo.Features.Copy} canMove={fsInfo.Features.Move} itemIdx={idx}>

                            </FileComponent>
                        </React.Fragment>
                    );
                }
                return null;
            });
        }
    };


    render() {
        const {isLoading, isDownloadProgress, downloadingItems, generatedLink, showLinkShareModal} = this.state;
        const {connectDropTarget, isOver, files, navigateUp, containerID, gridMode, canDrop} = this.props;
        const {remoteName} = this.props.currentPath;

        // console.log(this.props.searchQuery);

        if (isLoading || !files) {
            return (<div><Spinner color="primary"/> Loading</div>);
        } else {


            if (remoteName === "") {
                return (<div>No remote is selected. Select a remote from above to show files.</div>);
            }


            let dirComponentMap = this.getFileComponents(true);

            let fileComponentMap = this.getFileComponents(false);

            let renderElement = "";

            if (gridMode === "card") {

                renderElement = (

                    <Container fluid={true}>
                        <Row>
                            <UpButtonComponent upButtonHandle={() => navigateUp(containerID)} gridMode={gridMode}/>
                        </Row>
                        <Row>
                            <Col lg={3}>
                                <h3>Directories</h3>
                                <ScrollableDiv height={FILES_VIEW_HEIGHT}>
                                    {dirComponentMap}
                                </ScrollableDiv>
                            </Col>
                            <Col lg={9}>
                                <h3>Files</h3>
                                <ScrollableDiv height={FILES_VIEW_HEIGHT}>
                                    <Row>
                                        {fileComponentMap}
                                    </Row>
                                </ScrollableDiv>
                            </Col>

                        </Row>


                    </Container>

                )
            } else {


                renderElement = (


                    <Container fluid={true} className={"pd-0"}>
                        <ScrollableDiv height={FILES_VIEW_HEIGHT}>

                            <Table className="table-responsive-sm">
                                <thead>
                                <tr>
                                    <th className="d-none d-md-table-cell">x</th>
                                    <th>Name</th>
                                    <th>Size</th>
                                    <th className="d-none d-md-table-cell">Modified</th>
                                    <th>Actions</th>
                                </tr>
                                </thead>
                                <tbody>
                                <UpButtonComponent upButtonHandle={() => navigateUp(containerID)} gridMode={gridMode}/>
                                {files.length > 0 ? (
                                        <React.Fragment>
                                            <tr>
                                                <td colSpan={1} className="d-none d-md-block"/>
                                                <th colSpan={4}>Directories</th>
                                            </tr>
                                            {dirComponentMap}
                                            <tr>
                                                <td className="d-none d-md-table-cell"/>
                                                <th>Files</th>
                                                <td className="d-none d-md-table-cell"/>
                                                <td/>
                                                <td/>
                                            </tr>
                                            {fileComponentMap}
                                        </React.Fragment>
                                    ) :
                                    <tr>
                                        <td colSpan={1} className="d-none d-md-table-cell"/>
                                        <th colSpan={4}>Files</th>
                                    </tr>
                                }
                                </tbody>
                            </Table>
                        </ScrollableDiv>
                    </Container>


                );
            }


            return connectDropTarget(
                <div className={"row"}>
                    {isOver && canDrop && renderOverlay()}
                    <ErrorBoundary>

                        <Alert color="info" isOpen={isDownloadProgress} toggle={this.dismissAlert} sm={12}
                               lg={12}>
                            Downloading {downloadingItems} file(s). Please wait.
                        </Alert>

                        {renderElement}

                        <LinkShareModal closeModal={this.closeLinkShareModal} isVisible={showLinkShareModal}
                                        linkUrl={generatedLink}/>
                    </ErrorBoundary>
                </div>
            );
        }
    }

}

const propTypes = {
    containerID: PropTypes.string.isRequired,
    currentPath: PROP_CURRENT_PATH.isRequired,
    fsInfo: PROP_FS_INFO,
    gridMode: PropTypes.string,
    searchQuery: PropTypes.string,
    loadImages: PropTypes.bool.isRequired
};

const defaultProps = {
};


FilesView.propTypes = propTypes;
FilesView.defaultProps = defaultProps;


const mapStateToProps = (state, ownProps) => {
    const currentPath = state.explorer.currentPaths[ownProps.containerID];
    const visibilityFilter = state.explorer.visibilityFilters[ownProps.containerID];
    const gridMode = state.explorer.gridMode[ownProps.containerID];
    const searchQuery = state.explorer.searchQueries[ownProps.containerID];
    const loadImages = state.explorer.loadImages[ownProps.containerID];

    let fsInfo = {};
    const {remoteName, remotePath} = currentPath;

    if (currentPath && state.remote.configs) {

        const tempRemoteName = remoteName.split(':')[0];
        if (state.remote.configs[tempRemoteName])

            fsInfo = state.remote.configs[tempRemoteName];
    }

    const pathKey = `${remoteName}-${remotePath}`;

    let files = state.remote.files[pathKey];

    if (files) {
        files = files.files;
        // Filter according to visibility filters
        if (visibilityFilter) {
            files = changeListVisibility(files, visibilityFilter);
        }

        //Filter according to search query, if ny
        if (searchQuery) {
            files = changeSearchFilter(files, searchQuery);
        }
    }

    return {
        files,
        currentPath,
        fsInfo,
        gridMode,
        searchQuery,
        loadImages
    }
};

export default compose(
    connect(
        mapStateToProps, {getFiles, navigateUp, changePath}
    ),
    DropTarget(ItemTypes.FILECOMPONENT, filesTarget, collect)
)(FilesView)
