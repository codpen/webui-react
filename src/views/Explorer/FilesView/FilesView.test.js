import ReactDOM from "react-dom";
import React from "react";
import FilesView from "./FilesView";

it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<FilesView upButtonHandle={() => {
    }} remotePath={"local1"} updateRemotePathHandle={() => {
    }}/>, div);
    ReactDOM.unmountComponentAtNode(div);
});
