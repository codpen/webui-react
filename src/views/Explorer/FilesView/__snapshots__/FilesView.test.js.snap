// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Files View renders should match snapshot 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
